satellite/satellitedb/orders.go

```diff
@@ -214,3 +218,178 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
 	_, err = db.db.ExecContext(ctx, db.db.Rebind(statement), storageNodeID.Bytes(), serialNumber.Bytes())
 	return err
 }
+
+// ProcessOrders take a list of order requests and "settles" them in one transaction
+func (db *ordersDB) ProcessOrders(ctx context.Context, requests []*orders.ProcessOrderRequest) (responses []*pb.SettlementResponse, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if len(requests) == 0 {
+		return nil, err
+	}
+
+	tx, err := db.db.Begin()
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
+
+	now := time.Now().UTC()
+	intervalStart := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), 0, 0, 0, now.Location())
+
+	rejectedRequests := make(map[storj.SerialNumber]bool)
+	reject := func(serialNumber storj.SerialNumber) {
+		r := &pb.SettlementResponse{
+			SerialNumber: serialNumber,
+			Status:       pb.SettlementResponse_REJECTED,
+		}
+		rejectedRequests[serialNumber] = true
+		responses = append(responses, r)
+	}
+
+	// processes the insert to used serials table individually so we can handle
+	// the case where the order has already been processed.  Duplicates and previously
+	// processed orders are rejected
+	for _, request := range requests {
+		insert := "INSERT INTO used_serials (serial_number_id, storage_node_id) SELECT id, ? FROM serial_numbers WHERE serial_number = ?"
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(db.db.Rebind(insert), request.OrderLimit.StorageNodeId.Bytes(), request.OrderLimit.SerialNumber.Bytes())
+		if err != nil {
+			if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
+				reject(request.OrderLimit.SerialNumber)
+			} else {
+				return nil, Error.Wrap(err)
+			}
+		}
+	}
+
+	// call to get all the bucket IDs
+	query := db.buildGetBucketIdsQuery(len(requests))
+	statement := db.db.Rebind(query)
+
+	args := make([]interface{}, len(requests))
+	for i, request := range requests {
+		args[i] = request.OrderLimit.SerialNumber.Bytes()
+	}
+
+	rows, err := tx.Query(statement, args...)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	bucketMap := make(map[storj.SerialNumber][]byte)
+	for rows.Next() {
+		var serialNumber, bucketID []byte
+		err := rows.Scan(&serialNumber, &bucketID)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		sn, err := storj.SerialNumberFromBytes(serialNumber)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		bucketMap[sn] = bucketID
+	}
+
+	// build all the bandwidth updates into one sql statement
+	var updateRollupStatement string
```

Comment:
Is there a sql statement size limit?

Reply:
I ran benchmarks with batch sizes up to 5000.
sqlite fails at 1000, but that is because of the WHERE IN clause limits.  Not the size of sql statement.
I saw no issues with postgres.